Grant/contracts/BuilderGrant.sol

```diff
@@ -0,0 +1,334 @@
+
+        // Send Mini Buildings reward
+        if (amountForSelf > 0) {
+            season.rankers[ranking].claimedAmount += amountForSelf;
+            season.totalClaimed += amountForSelf;
+            if (!_mintBuildings(amountForSelf, msgSender)) revert MintBuildingsFailed();
+        }
+
+        // Set donations
+        if (amountForDonation > 0) {
+            for (uint256 i = 3; i < amountForDonation + 3; ++i) {
+                season.rankers[i].donationReceived[ranking] = true;
+            }
+        }
+
+        emit ClaimByTop3(msgSender, seasonId, ranking, claimType, amountForSelf, amountForDonation);
+    }
+
+    function isDonationClaimableNow(uint256 seasonId) public view validSeasonId(seasonId) returns (bool) {
+        Season storage season = seasons[seasonId];
+
+        // If the donation claim deadline is passed, return false
+        if (season.claimStartedAt + DONATION_CLAIM_DEADLINE < block.timestamp) revert DonationClaimDeadlineReached();
+
+        // If the top 3 claim deadline is passed, return true
+        if (season.claimStartedAt + TOP3_CLAIM_DEADLINE < block.timestamp) return true;
+
+        // Even if the top 3 claim deadline is NOT passed, if all top 3 have already been decided, return true
+        if (season.grants[0].claimedType > 0 && season.grants[1].claimedType > 0 && season.grants[2].claimedType > 0)
+            return true;
+
+        return false;
+    }
+
+    function claimableDonationAmount(
+        uint256 seasonId,
+        uint256 ranking
+    ) public view validSeasonId(seasonId) returns (uint16 claimableAmount) {
+        if (!isDonationClaimableNow(seasonId)) revert DonationNotClaimableYet();
+
+        Season storage season = seasons[seasonId];
+        if (season.rankers.length <= ranking) revert InvalidRankingParam();
+        if (season.rankers[ranking].claimedAmount > 0) revert DonationAlreadyClaimed();
+
+        bool[3] storage donationReceived = season.rankers[ranking].donationReceived;
+
+        for (uint256 i = 0; i < 3; ++i) {
+            if (donationReceived[i] == true) {
+                ++claimableAmount;
+            }
+        }
+    }
+
+    function claimDonation(uint256 seasonId, uint256 ranking) external validSeasonId(seasonId) {
+        uint16 claimableAmount = claimableDonationAmount(seasonId, ranking);
+        if (claimableAmount == 0) revert NoDonationToClaim();
+
+        Season storage season = seasons[seasonId];
+        address msgSender = _msgSender();
+        if (season.rankers[ranking].wallet != msgSender) revert PermissionDenied();
+
+        season.rankers[ranking].claimedAmount += claimableAmount;
+        season.totalClaimed += claimableAmount;
+        if (!_mintBuildings(claimableAmount, msgSender)) revert MintBuildingsFailed();
+
+        emit ClaimDonation(msgSender, seasonId, ranking, claimableAmount);
+    }
+
+    function _mintBuildings(uint256 count, address to) private returns (bool) {
```

Comment:
you probably should use reentrancy guard on this, since its making a call to an external contract

https://docs.openzeppelin.com/contracts/5.x/api/utils#ReentrancyGuard

Reply:
try catch should be used also here

```
function _mintBuildings(uint256 count, address to) private nonReentrant returns (bool) {
    try BOND.mint(MINI_BUILDING_ADDRESS, count, HUNT_PER_MINI_BUILDING * count, to) {
        return true;
    } catch {
        return false;
    }
}
```


then wherever its used pretty sure you can do 

```
 if (!_mintBuildings(claimableAmount, msgSender)) revert MintBuildingsFailed();
```

Reply:
I prefer using the CEI pattern over the `nonReentrant` library as long as the function is simple enough because `nonReentrant` adds extra gas cost, even though it's small.

Reply:
+1 to CEI pattern

Reply:
@stuckinaboot You're @aspyn on Warpcast, right? Thanks for reviewing and please leave your public wallet address here for the bounty :)